server/routerlicious/packages/services/src/mongodb.ts

```diff
@@ -7,103 +7,172 @@ import { assert } from "console";
+        const req = async () => this.updateCore(filter, set, addToSet, false);
+        return this.requestWithRetry(
+            req, // request
+            "MongoCollection.update", // callerName
+        );
     }
 
     // eslint-disable-next-line @typescript-eslint/ban-types
     public async updateMany(filter: object, set: any, addToSet: any): Promise<void> {
-        return this.updateManyCore(filter, set, addToSet, false);
+        const req = async () => this.updateManyCore(filter, set, addToSet, false);
+        return this.requestWithRetry(
+            req, // request
+            "MongoCollection.updateMany", // callerName
+        );
     }
 
     // eslint-disable-next-line @typescript-eslint/ban-types
     public async upsert(filter: object, set: any, addToSet: any): Promise<void> {
-        return this.updateCore(filter, set, addToSet, true);
+        const req = async () => this.updateCore(filter, set, addToSet, true);
+        return this.requestWithRetry(
+            req, // request
+            "MongoCollection.upsert", // callerName
+        );
     }
 
     public async distinct(key: any, query: any): Promise<any> {
-        return this.collection.distinct(key, query);
+        const req = async () => this.collection.distinct(key, query);
+        return this.requestWithRetry(
+            req, // request
+            "MongoCollection.distinct", // callerName
+        );
     }
 
     public async deleteOne(filter: any): Promise<any> {
-        return this.collection.deleteOne(filter);
+        const req = async () => this.collection.deleteOne(filter);
+        return this.requestWithRetry(
+            req, // request
+            "MongoCollection.deleteOne", // callerName
+        );
     }
 
     public async deleteMany(filter: any): Promise<any> {
-        return this.collection.deleteMany(filter);
+        const req = async () => this.collection.deleteMany(filter);
+        return this.requestWithRetry(
+            req, // request
+            "MongoCollection.deleteMany", // callerName
+        );
     }
 
     public async insertOne(value: T): Promise<any> {
-        const result = await this.collection.insertOne(value);
-        return result.insertedId;
+        const req = async () => {
+            const result = await this.collection.insertOne(value);
+            return result.insertedId;
+        };
+        return this.requestWithRetry(
+            req, // request
+            "MongoCollection.insertOne", // callerName
+        );
     }
 
     public async insertMany(values: T[], ordered: boolean): Promise<void> {
-        await this.collection.insertMany(values, { ordered: false });
+        const req = async () => this.collection.insertMany(values, { ordered: false });
+        await this.requestWithRetry(
+            req, // request
+            "MongoCollection.insertMany", // callerName
+        );
     }
 
     public async createIndex(index: any, unique: boolean): Promise<void> {
```

Comment:
>createIndex [](http://example.com/codeflow?start=17&length=11)

(educate me) why do we use a different pattern for createIndex?

Reply:
For 2 reasons:
1. createIndex was wrapped under `try/catch` and never bubbled up. So with the retry logic added on, we don't want to bubble up the exception as well.
<img width="1358" alt="image" src="https://user-images.githubusercontent.com/3719246/201681435-b185bc29-0386-4cf6-9e8a-73e30d25659b.png">

2. Create indexes with unique restriction were always failing in our system but no one found this before. The error will continue without a DB rebuild, which is a hard and long work.Also create index mostly happened at service bootstrap time. If we bubble up at that time, service will failed to start. Which I guess people did the `try/catch` and log error in the 1st place?

So we can't bubble up exceptions like others

Reply:
Makes sense. Thanks for explaining! Add some comments for this?

Reply:
Added comments in 4b800a9eb65d61a7a97c285fe895ce48f7662785